File: ele.cpp
#include <iostream>
#include <cstdlib>
#include <string>
#include <thread>
#include <chrono>

struct Elevador {
    int nivel;
    bool niveles[5];
    const char* nombre;
};

Elevador elevador1 = {
    1, {false, true, false, false, false}, "Elevador 1"
};

Elevador elevador2 = {
    2, {false, false, true, false, false}, "Elevador 2"
};

Elevador* elevadores[] = { &elevador1, &elevador2 };
const int numElevadores = 2;

void imprimirNiveles(const Elevador* elevador) {
    std::cout << "\n🏗️ Estado de " << elevador->nombre << "\n";
    for (int i = 4; i >= 0; i--) {
        std::cout << "Nivel " << i << ": "
                  << (elevador->nivel == i ? "🟢" : "⚪") << "\n";
    }
}

Elevador* buscarElevador(Elevador* elevadores[], int numElevadores, int nivelDestino) {
    Elevador* mejorElevador = elevadores[0];
    int menorDistancia = std::abs(elevadores[0]->nivel - nivelDestino);

    for (int i = 1; i < numElevadores; i++) {
        int distancia = std::abs(elevadores[i]->nivel - nivelDestino);
        if (distancia < menorDistancia) {
            menorDistancia = distancia;
            mejorElevador = elevadores[i];
        }
    }
    return mejorElevador;
}

void simularMovimiento(Elevador* elevador, int nivelDestino) {
    if (nivelDestino < 0 || nivelDestino > 4) {
        std::cout << "❌ Nivel destino inválido. Debe estar entre 0 y 4.\n";
        return;
    }

    std::cout << "\n🚀 " << elevador->nombre << " se mueve del nivel "
              << elevador->nivel << " al nivel " << nivelDestino << "\n";

    const char* direccion = (nivelDestino > elevador->nivel) ? "⬆️" : "⬇️";
    int pasos = std::abs(nivelDestino - elevador->nivel);

    std::cout << direccion << " Distancia: " << pasos << " niveles\n";

    elevador->nivel = nivelDestino;
    for (int i = 0; i < 5; i++) {
        elevador->niveles[i] = (i == nivelDestino);
    }

    std::cout << "✅ " << elevador->nombre << " llegó al nivel " << nivelDestino << "\n";
    imprimirNiveles(elevador);
}

int main() {
    std::this_thread::sleep_for(std::chrono::milliseconds(1000));

    std::cout << "🏢 Sistema de Elevadores - Simulación\n";
    std::cout << "=====================================\n";

    std::cout << "\n📊 Estado inicial:\n";
    for (int i = 0; i < numElevadores; i++) {
        std::cout << elevadores[i]->nombre << ": Nivel " << elevadores[i]->nivel << "\n";
        imprimirNiveles(elevadores[i]);
    }

    int nivelSolicitado = 3;
    std::cout << "\n🎯 Solicitud para nivel " << nivelSolicitado << "\n";

    Elevador* elevadorSeleccionado = buscarElevador(elevadores, numElevadores, nivelSolicitado);

    std::cout << "🏆 Elevador seleccionado: " << elevadorSeleccionado->nombre << "\n";

    simularMovimiento(elevadorSeleccionado, nivelSolicitado);

    std::cout << "\n📊 Estado final:\n";
    for (int i = 0; i < numElevadores; i++) {
        std::cout << elevadores[i]->nombre << ": Nivel " << elevadores[i]->nivel << "\n";
        imprimirNiveles(elevadores[i]);
    }

    return 0;
}

File: ele.ts
interface Elevador {
  nivel: number;
  niveles: boolean[];
  nombre: string;
}

function posicionElevador(niveles: boolean[]): number {
  const nivel = niveles.findIndex((nivel) => nivel === true);
  return nivel;
}

const elevador1: Elevador = {
  nivel: posicionElevador([false, true, false, false, false]), // Nivel 1
  niveles: [false, true, false, false, false],
  nombre: "Elevador 1",
};

const elevador2: Elevador = {
  nivel: posicionElevador([false, false, true, false, false]), // Nivel 2
  niveles: [false, false, true, false, false],
  nombre: "Elevador 2",
};

const elevadores = [elevador1, elevador2];

function buscarElevador(
  elevadores: Elevador[],
  nivelDestino: number,
): Elevador {
  let mejorElevador = elevadores[0];
  let menorDistancia = Math.abs(elevadores[0].nivel - nivelDestino);

  for (let i = 1; i < elevadores.length; i++) {
    const distancia = Math.abs(elevadores[i].nivel - nivelDestino);
    if (distancia < menorDistancia) {
      menorDistancia = distancia;
      mejorElevador = elevadores[i];
    }
  }

  return mejorElevador;
}

// Función para simular el movimiento
function simularMovimiento(elevador: Elevador, nivelDestino: number): void {
  console.log(
    `\n🚀 ${elevador.nombre} se mueve del nivel ${elevador.nivel} al nivel ${nivelDestino}`,
  );

  const direccion = nivelDestino > elevador.nivel ? "⬆️" : "⬇️";
  const pasos = Math.abs(nivelDestino - elevador.nivel);

  console.log(`${direccion} Distancia: ${pasos} niveles`);

  // Actualizar posición del elevador
  elevador.nivel = nivelDestino;
  elevador.niveles = elevador.niveles.map((_, index) => index === nivelDestino);

  console.log(`✅ ${elevador.nombre} llegó al nivel ${nivelDestino}`);
}

// Pruebas del algoritmo
console.log("🏢 Sistema de Elevadores - Simulación");
console.log("=====================================");

console.log("\n📊 Estado inicial:");
elevadores.forEach((elevador) => {
  console.log(`${elevador.nombre}: Nivel ${elevador.nivel}`);
});

// Buscar elevador para nivel 3
const nivelSolicitado = 3;
const elevadorSeleccionado = buscarElevador(elevadores, nivelSolicitado);

console.log(`\n🎯 Solicitud para nivel ${nivelSolicitado}`);
console.log(`🏆 Elevador seleccionado: ${elevadorSeleccionado.nombre}`);

// Simular el movimiento
simularMovimiento(elevadorSeleccionado, nivelSolicitado);

console.log("\n📊 Estado final:");
elevadores.forEach((elevador) => {
  console.log(`${elevador.nombre}: Nivel ${elevador.nivel}`);
});
